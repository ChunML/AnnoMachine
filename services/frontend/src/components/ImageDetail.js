import React, { useState, useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import BoxesDetail from './BoxesDetail';
import UploadInfo from './UploadInfo';
import ImageAnnoDisplay from './ImageAnnoDisplay';

function ImageDetail({ image }) {
  const [drawBoxes, setDrawBoxes] = useState([]);
  const [boxes, setBoxes] = useState([]);
  const [editModes, setEditModes] = useState([]);
  const [addBoxMode, setAddBoxMode] = useState(false);
  const [scale, setScale] = useState(1);
  const [svgWidth, setSvgWidth] = useState(0);
  const [svgHeight, setSvgHeight] = useState(0);
  const ref = useRef(null);

  useEffect(() => {
    if (ref.current && ref.current.offsetWidth < image.width) {
      setScale(ref.current.offsetWidth / image.width);
      setSvgWidth(ref.current.offsetWidth);
      setSvgHeight(scale * image.height);
    } else {
      setSvgWidth(image.width);
      setSvgHeight(image.height);
    }
    if (image.boxes) {
      setBoxes(image.boxes.sort((a, b) => a.id - b.id));
      setEditModes(image.boxes.map(box => false));
    }
  }, [image, scale]);

  if (image.length === 0) {
    return <div>Loading...</div>
  } else {
    image = image[0];
  }

  const onEyeIconClick = box => {
    const currentBox = drawBoxes.filter(drawBox => drawBox.id === box.id);
    if (currentBox.length === 0) {
      setDrawBoxes([...drawBoxes, box]);
    } else {
      const otherBoxes = drawBoxes.filter(drawBox => drawBox.id !== box.id);
      setDrawBoxes([...otherBoxes]);
    }
  }

  const onInputChange = box => {
    updateBoxes(box);
  }

  const onCheckIconClick = box => {
    if (box.id === Math.max(...boxes.map(box => box.id))) {
      setAddBoxMode(false);
    }
    updateBoxes(box);
  }

  const updateBoxes = box => {
    const otherBoxes = boxes.filter(originalBox => originalBox.id !== box.id);
    const newBoxes = [ ...otherBoxes, box ];
    setBoxes(newBoxes.sort((a, b) => a.id - b.id));
    setDrawBoxes(drawBoxes.map(drawBox => {
      if (drawBox.id === box.id) {
        return box;
      }
      return drawBox;
    }));
  }

  const onAddBoxButtonClick = () => {
    setAddBoxMode(true);
    const newBox = {
      id: Math.max(...boxes.map(box => box.id)) + 1,
      label: '',
      x_min: 0,
      y_min: 0,
      x_max: 10,
      y_max: 10
    };
    setBoxes([ ...boxes, newBox ]);
    setDrawBoxes([ ...drawBoxes, newBox ]);
    setEditModes([ ...editModes, true ]);
  }

  const onUndoBoxAddingButtonClick = () => {
    setAddBoxMode(false);
    const oldBoxes = boxes.filter(
      box => box.id < Math.max(...boxes.map(box => box.id)));
    setBoxes([ ...oldBoxes ]);
    const oldDrawBoxes = drawBoxes.filter(
      box => box.id < Math.max(...boxes.map(box => box.id)));
    setDrawBoxes([ ...oldDrawBoxes ]);
    setEditModes([ ...oldBoxes.map(box => false) ]);
  }

  return (
    <div className="ui center aligned stackable two column grid">
      <div className="row">
        <div className="column" ref={ ref }>
          <ImageAnnoDisplay
            svgWidth={ svgWidth || 0 }
            svgHeight={ svgHeight || 0 }
            imageWidth={ image.width }
            imageHeight={ image.height }
            scale={ scale }
            drawBoxes={ drawBoxes }
            name={ image.name }
          />
          <p></p>
          <a href={`${process.env.REACT_APP_API_URL}/api/annotations/kitti/${image.name.replace('.jpg', '.txt')}`}>
            <button className="ui primary button">Download annotation</button>
          </a>
        </div>
        <div className="column">
          <UploadInfo username={ image.user.username } uploaded_at={ image.uploaded_at } />
          <BoxesDetail
            boxes={ boxes || [] }
            editModes= { editModes || [] }
            onEyeIconClick={ onEyeIconClick }
            onCheckIconClick={ onCheckIconClick }
            onInputChange={ onInputChange }
          />
          { addBoxMode ? (
            <button
              className='ui red circular icon button'
              onClick={ onUndoBoxAddingButtonClick }
            >
              <i className='trash alternate outline icon'></i>
            </button>) : (
            <button
              className='ui circular primary button'
              onClick={ onAddBoxButtonClick }
            >
              Add
            </button>)}
        </div>
      </div>
    </div>
  );
}

ImageDetail.propTypes = {
  image: PropTypes.array.isRequired
};

export default ImageDetail;