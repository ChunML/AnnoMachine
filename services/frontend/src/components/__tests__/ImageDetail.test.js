import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import ImageDetail from '../ImageDetail';
import '../../setupTests';

const image = {
  name: 'testImage',
  uploaded_at: '2000/01/01 01:02',
  user: {
    username: 'testUser',
  },
  width: 100,
  height: 100,
  boxes: [
    {
      id: 1,
      label: 'dog',
      x_min: 10,
      y_min: 10,
      x_max: 20,
      y_max: 20,
    },
    {
      id: 2,
      label: 'cat',
      x_min: 20,
      y_min: 20,
      x_max: 40,
      y_max: 40,
    },
  ],
};

describe('No image is passed', () => {
  it('ImageDetail renders properly', () => {
    const wrapper = shallow(<ImageDetail createMessage={jest.fn()} />);
    expect(wrapper.text()).toEqual('Loading...');
  });

  it('ImageDetail renders a snapshot properly', () => {
    const tree = renderer
      .create(<ImageDetail createMessage={jest.fn()} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('One image is properly passed', () => {
  it('ImageDetail renders properly', () => {
    const wrapper = shallow(
      <ImageDetail image={image} createMessage={jest.fn()} />
    );
    const columns = wrapper.find('.row > .half-width-item');
    expect(columns.length).toBe(2);
    expect(columns.at(0).find('ImageAnnoDisplay')).toHaveLength(1);
    expect(columns.at(0).find('.button')).toHaveLength(1);
    expect(
      columns
        .at(0)
        .find('.button')
        .text()
    ).toEqual('Download annotation');
    expect(columns.at(1).find('UploadInfo')).toHaveLength(1);
    expect(columns.at(1).find('BoxesDetail')).toHaveLength(1);
    expect(columns.at(1).find('.button')).toHaveLength(1);
    expect(
      columns
        .at(1)
        .find('.button')
        .get(0).props.children
    ).toContain('Add');
    columns
      .at(1)
      .find('.button')
      .simulate('click');
    expect(
      wrapper
        .find('.row > .half-width-item')
        .at(1)
        .find('.button')
        .get(0).props.children
    ).toContain('Undo');
  });

  it('ImageDetail renders a snapshot properly', () => {
    const tree = renderer
      .create(<ImageDetail image={image} createMessage={jest.fn()} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
