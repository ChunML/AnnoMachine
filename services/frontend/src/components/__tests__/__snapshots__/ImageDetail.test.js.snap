// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageDetail renders a snapshot properly 1`] = `
<div
  className="ui center aligned stackable two column grid"
>
  <div
    className="row"
  >
    <div
      className="column"
    >
      <svg
        height={0}
        style={
          Object {
            "marginLeft": "-14px",
          }
        }
        width={0}
        xmlns="http://www.w3.org/2000/svg"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      >
        <image
          height={0}
          style={
            Object {
              "height": "100%",
            }
          }
          width={0}
          x="0"
          xlinkHref="http://localhost/api/uploads/testImage"
          y="0"
        />
      </svg>
      <p />
      <a
        href="http://localhost/api/annotations/kitti/testImage"
      >
        <button
          className="ui primary button"
        >
          Download annotation
        </button>
      </a>
    </div>
    <div
      className="column"
    >
      <div
        className="ui left aligned red segment"
      >
        <p>
          Author: 
          <span
            style={
              Object {
                "color": "tomato",
                "fontSize": "1.6rem",
              }
            }
          >
            testUser
          </span>
        </p>
        <p>
          Uploaded at: 
          <span
            style={
              Object {
                "color": "chocolate",
                "fontStyle": "italic",
              }
            }
          >
            2000/01/01 01:02
          </span>
        </p>
      </div>
      <div
        className="ui segments"
      />
      <button
        className="ui circular primary button"
        onClick={[Function]}
      >
        Add
      </button>
    </div>
  </div>
</div>
`;
